R/epiparameter.R

```diff
@@ -0,0 +1,996 @@
+#'     journal = "Epi Journal",
+#'     doi = "10.19832/j.1366-9516.2012.09147.x"
+#'   ),
+#'   metadata = create_metadata(
+#'     units = "days",
+#'     sample_size = 10,
+#'     region = "UK",
+#'     transmission_mode = "natural_human_to_human",
+#'     inference_method = "MLE"
+#'   ),
+#'   method_assess = create_method_assess(
+#'     censored = TRUE
+#'   ),
+#'   notes = "No notes"
+#' )
+epiparameter <- function(disease,
+                         pathogen = NA_character_,
+                         epi_name,
+                         prob_distribution = create_prob_distribution(
+                           prob_distribution = NA_character_
+                         ),
+                         uncertainty = create_uncertainty(),
+                         summary_stats = create_summary_stats(),
+                         citation = create_citation(),
+                         metadata = create_metadata(),
+                         method_assess = create_method_assess(),
+                         notes = NULL,
+                         auto_calc_params = TRUE,
+                         ...) {
+  # check input
+  checkmate::assert_string(disease)
+  checkmate::assert_character(pathogen)
+  checkmate::assert_string(epi_name)
+  stopifnot(
+    "Probability distribution must be a distribution object or a character" =
+      !inherits(prob_distribution, c("distribution", "distcrete")) ||
+      !is.character(prob_distribution)
+  )
+  checkmate::assert_list(uncertainty, names = "unique")
+  checkmate::assert_list(
+    summary_stats,
+    types = c("numeric", "logical"),
+    names = "unique",
+    null.ok = TRUE
+  )
+  checkmate::assert_class(citation, classes = "bibentry")
+  checkmate::assert_list(metadata)
+  checkmate::assert_list(method_assess)
+  checkmate::assert_character(notes, null.ok = TRUE)
+
+  # call epiparameter constructor
+  epiparameter <- new_epiparameter(
+    disease = disease,
+    pathogen = pathogen,
+    epi_name = epi_name,
+    prob_distribution = prob_distribution,
+    uncertainty = uncertainty,
+    summary_stats = summary_stats,
+    auto_calc_params = auto_calc_params,
+    citation = citation,
+    metadata = metadata,
+    method_assess = method_assess,
+    notes = notes,
+    ...
+  )
+
+  # uncertainty is checked after new_epiparameter to use <epiparameter> methods
+  epiparameter$uncertainty <- .clean_uncertainty(
+    epiparameter,
+    prob_distribution = prob_distribution,
+    uncertainty_missing = missing(uncertainty)
+  )
+
+  param_names <- names(get_parameters(epiparameter)) %||% NA_character_
+
+  stopifnot(
+    "uncertainty must be provided for each parameter" =
+      length(param_names) == length(epiparameter$uncertainty),
+    "parameters and uncertainty must be named and match" =
+      identical(param_names, names(epiparameter$uncertainty)) ||
+      is.na(param_names)
+  )
+
+  # call epiparameter validator
+  assert_epiparameter(epiparameter)
```

Comment:
I agree this is superfluous. I like having it to ensure the object created is valid. I think it is impossible to get to the `assert_epiparameter()` call with an invalid `<epiparameter>` object, but just as an extra layer of defence I've added it. 

As the time to run this extra function is negligible I do not think there is much reason to remove, but if users were creating many thousand/million `<epiparameter>` objects this would be a good first thing to remove for a speed gain.

If there are other reasons to remove that I've overlooked please let me know.